beacon_node/network/src/sync/manager.rs

```diff
@@ -0,0 +1,725 @@
+        if remote.head_slot.sub(local.head_slot).as_usize() < SLOT_IMPORT_TOLERANCE {
+            trace!(self.log, "Ignoring full sync with peer";
+            "peer" => format!("{:?}", peer_id),
+            "peer_head_slot" => remote.head_slot,
+            "local_head_slot" => local.head_slot,
+            );
+            // remove the peer from the queue if it exists
+            self.import_queue.remove(&peer_id);
+            return;
+        }
+
+        if let Some(block_requests) = self.import_queue.get_mut(&peer_id) {
+            // update the target head slot
+            if remote.head_slot > block_requests.target_head_slot {
+                block_requests.target_head_slot = remote.head_slot;
+            }
+        } else {
+            let block_requests = BlockRequests {
+                target_head_slot: remote.head_slot, // this should be larger than the current head. It is checked in the SyncManager before add_peer is called
+                target_head_root: remote.head_root,
+                downloaded_blocks: Vec::new(),
+                state: BlockRequestsState::Queued,
+                forward_sync: true,
+                current_start_slot: self.chain.best_slot(),
+            };
+            self.import_queue.insert(peer_id, block_requests);
+        }
+    }
+
+    pub fn beacon_blocks_response(
+        &mut self,
+        peer_id: PeerId,
+        request_id: RequestId,
+        mut blocks: Vec<BeaconBlock<T::EthSpec>>,
+    ) {
+        // find the request
+        let block_requests = match self
+            .import_queue
+            .get_mut(&peer_id)
+            .filter(|r| r.state == BlockRequestsState::Pending(request_id))
+        {
+            Some(req) => req,
+            _ => {
+                // No pending request, invalid request_id or coding error
+                warn!(self.log, "BeaconBlocks response unknown"; "request_id" => request_id);
+                return;
+            }
+        };
+
+        // If we are syncing up to a target head block, at least the target head block should be
+        // returned. If we are syncing back to our last finalized block the request should return
+        // at least the last block we received (last known block). In diagram form:
+        //
+        //     unknown blocks       requested blocks        downloaded blocks
+        // |-------------------|------------------------|------------------------|
+        // ^finalized slot     ^ requested start slot   ^ last known block       ^ remote head
+
+        if blocks.is_empty() {
+            debug!(self.log, "BeaconBlocks response was empty"; "request_id" => request_id);
+            block_requests.update_start_slot();
+            return;
+        }
+
+        // verify the range of received blocks
+        // Note that the order of blocks is verified in block processing
+        let last_sent_slot = blocks[blocks.len() - 1].slot;
```

Comment:
It seems like blocks could be be empty?

https://github.com/sigp/lighthouse/blob/1bea1755c46d17fff9c6cea56e55691bf5cfa1b9/beacon_node/network/src/message_handler.rs#L306-L308

Reply:
I thought this wasn't possible. Will check and get back to this

Reply:
See line 168. This condition is checked.